File: macos/Tests/TranskritoTests/CorrectionEngineTests.swift
import Foundation
import Testing
@testable import Transkrito

/// Runs the shared vectors in shared/correction-tests.json — the same file the Windows tests use.
struct CorrectionEngineTests {
    static let vectors: [String: Any] = {
        let url = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
            .appendingPathComponent("shared/correction-tests.json")
        let data = try! Data(contentsOf: url)
        return try! JSONSerialization.jsonObject(with: data) as! [String: Any]
    }()

    static let iso: ISO8601DateFormatter = { let f = ISO8601DateFormatter(); f.formatOptions = [.withInternetDateTime]; return f }()

    static func entry(_ d: [String: Any]) -> DictionaryEntry {
        DictionaryEntry(
            type: d["type"] as! String,
            text: d["text"] as? String,
            hear: d["hear"] as? String,
            write: d["write"] as? String,
            created: (d["created"] as? String).flatMap { iso.date(from: $0) } ?? Date(timeIntervalSince1970: 0)
        )
    }

    static var cases: [[String: Any]] { vectors["cases"] as! [[String: Any]] }
    static var warningCases: [[String: Any]] { vectors["warnings"] as! [[String: Any]] }
    static var biasCases: [[String: Any]] { vectors["bias"] as! [[String: Any]] }

    @Test(arguments: CorrectionEngineTests.cases.map { $0["name"] as! String })
    func applyMatchesSharedVectors(name: String) throws {
        let c = Self.cases.first { $0["name"] as! String == name }!
        let entries = (c["entries"] as! [[String: Any]]).map(Self.entry)
        let input = c["input"] as! String
        let expected = c["expected"] as! String
        let expectedEvents = (c["events"] as! [[String: Any]]).map {
            CorrectionEvent(hear: $0["hear"] as! String, write: $0["write"] as! String, matched: $0["matched"] as! String, count: $0["count"] as! Int)
        }

        let (text, events) = CorrectionEngine.apply(input, entries: entries)
        #expect(text == expected)
        #expect(events == expectedEvents)

        // Idempotence: a second pass changes nothing.
        let (again, againEvents) = CorrectionEngine.apply(text, entries: entries)
        #expect(again == text)
        #expect(againEvents.isEmpty)
    }

    @Test(arguments: CorrectionEngineTests.warningCases.map { $0["name"] as! String })
    func warningsMatchSharedVectors(name: String) {
        let c = Self.warningCases.first { $0["name"] as! String == name }!
        let others = (c["others"] as! [[String: Any]]).map(Self.entry)
        let expect = (c["expect"] as! [String]).sorted()
        let got = Array(Set(DictionaryWarnings.check(hear: c["hear"] as! String, write: c["write"] as? String, others: others)
            .map { $0.kind.rawValue })).sorted()
        #expect(got == expect)
    }

    @Test func biasTermsMatchSharedVectors() {
        for c in Self.biasCases {
            let entries = (c["entries"] as! [[String: Any]]).map(Self.entry)
            let expect = c["expect"] as! [String]
            #expect(Bias.terms(entries, maxTerms: c["maxTerms"] as! Int) == expect)
        }
    }
}

File: windows/Transkrito.Tests/CorrectionEngineTests.cs
using System.Text.Json;
using Transkrito.Dictionary;
using Xunit;

namespace Transkrito.Tests;

/// <summary>Runs the shared vectors in shared/correction-tests.json (copied next to the test binary).</summary>
public class CorrectionEngineTests
{
    private static readonly JsonDocument Vectors = JsonDocument.Parse(
        File.ReadAllText(Path.Combine(AppContext.BaseDirectory, "correction-tests.json")));

    private static DictionaryEntry Entry(JsonElement e) => new()
    {
        Type = e.GetProperty("type").GetString()!,
        Text = e.TryGetProperty("text", out var t) ? t.GetString() : null,
        Hear = e.TryGetProperty("hear", out var h) ? h.GetString() : null,
        Write = e.TryGetProperty("write", out var w) ? w.GetString() : null,
        Created = e.TryGetProperty("created", out var c) ? c.GetDateTimeOffset() : DateTimeOffset.UnixEpoch,
    };

    public static IEnumerable<object[]> Cases() =>
        Vectors.RootElement.GetProperty("cases").EnumerateArray().Select(c => new object[] { c.GetProperty("name").GetString()! });

    private static JsonElement Find(string section, string name) =>
        Vectors.RootElement.GetProperty(section).EnumerateArray().First(c => c.GetProperty("name").GetString() == name);

    [Theory]
    [MemberData(nameof(Cases))]
    public void Apply_matches_shared_vectors(string name)
    {
        var c = Find("cases", name);
        var entries = c.GetProperty("entries").EnumerateArray().Select(Entry).ToList();
        var input = c.GetProperty("input").GetString()!;
        var expected = c.GetProperty("expected").GetString()!;

        var (text, events) = CorrectionEngine.Apply(input, entries);

        Assert.Equal(expected, text);
        var expectedEvents = c.GetProperty("events").EnumerateArray().Select(e => new CorrectionEvent(
            e.GetProperty("hear").GetString()!, e.GetProperty("write").GetString()!,
            e.GetProperty("matched").GetString()!, e.GetProperty("count").GetInt32())).ToList();
        Assert.Equal(expectedEvents, events);

        // Idempotence: a second pass changes nothing.
        var (again, againEvents) = CorrectionEngine.Apply(text, entries);
        Assert.Equal(text, again);
        Assert.Empty(againEvents);
    }

    public static IEnumerable<object[]> WarningCases() =>
        Vectors.RootElement.GetProperty("warnings").EnumerateArray().Select(c => new object[] { c.GetProperty("name").GetString()! });

    [Theory]
    [MemberData(nameof(WarningCases))]
    public void Warnings_match_shared_vectors(string name)
    {
        var c = Find("warnings", name);
        var hear = c.GetProperty("hear").GetString()!;
        var write = c.TryGetProperty("write", out var w) ? w.GetString() : null;
        var others = c.GetProperty("others").EnumerateArray().Select(Entry).ToList();
        var expect = c.GetProperty("expect").EnumerateArray().Select(e => e.GetString()!).OrderBy(x => x).ToList();

        var got = DictionaryWarnings.Check(hear, write, others)
            .Select(x => x.Kind.ToString().ToLowerInvariant()).Distinct().OrderBy(x => x).ToList();

        Assert.Equal(expect, got);
    }

    [Fact]
    public void Bias_terms_match_shared_vectors()
    {
        foreach (var c in Vectors.RootElement.GetProperty("bias").EnumerateArray())
        {
            var entries = c.GetProperty("entries").EnumerateArray().Select(Entry).ToList();
            var max = c.GetProperty("maxTerms").GetInt32();
            var expect = c.GetProperty("expect").EnumerateArray().Select(e => e.GetString()!).ToList();
            Assert.Equal(expect, Bias.Terms(entries, max));
        }
    }
}
